vm/src/value.cpp

```diff
@@ -0,0 +1,533 @@
+{
+}
+
+Value::Value(std::vector<Vector_3> vs)
+    : type_(Type::vector_3_array),
+      vector_3_array_(std::shared_ptr<const std::vector<Vector_3>>(
+          new std::vector<Vector_3>(std::move(vs))))
+{
+}
+
+Value::Value(std::vector<Matrix_3> ms)
+    : type_(Type::matrix_3_array),
+      matrix_3_array_(std::shared_ptr<const std::vector<Matrix_3>>(
+          new std::vector<Matrix_3>(std::move(ms))))
+{
+}
+
+Value::Value(std::vector<Matrix_4> ms)
+    : type_(Type::matrix_4_array),
+      matrix_4_array_(std::shared_ptr<const std::vector<Matrix_4>>(
+          new std::vector<Matrix_4>(std::move(ms))))
+{
+}
+
+Value::Value(const Value& x)
+{
+    construct(x);
+}
+
+Value::Value(Value&& x)
+{
+    construct(x);
+}
+
+Value&
+Value::operator=(const Value& x)
+{
+    if (this != &x) {
+        destroy();
+        construct(x);
+    }
+    return *this;
+}
+
+Value&
+Value::operator=(Value&& x)
+{
+    if (this != &x) {
+        destroy();
+        construct(x);
+    }
+    return *this;
+}
+
+Value::~Value()
+{
+    destroy();
+}
+
+Type
+Value::type() const
+{
+    return type_;
+}
+
+const bool&
+Value::boolean() const
+{
+    assert(type_ == Type::boolean);
+    return boolean_;
+}
+
+const double&
+Value::number() const
+{
+    assert(type_ == Type::number);
+    return number_;
+}
+
+const Vector_2&
+Value::vector_2() const
+{
+    assert(type_ == Type::vector_2);
+    return *vector_2_;
+}
+
+const Vector_3&
+Value::vector_3() const
+{
+    assert(type_ == Type::vector_3);
+    return *vector_3_;
+}
+
+const Matrix_3&
+Value::matrix_3() const
+{
+    assert(type_ == Type::matrix_3);
+    return *matrix_3_;
+}
+
+const Matrix_4&
+Value::matrix_4() const
+{
+    assert(type_ == Type::matrix_4);
+    return *matrix_4_;
+}
+
+const std::vector<bool>&
+Value::boolean_array() const
+{
+    assert(type_ == Type::boolean_array);
+    return *boolean_array_;
+}
+
+const std::vector<double>&
+Value::number_array() const
+{
+    assert(type_ == Type::number_array);
+    return *number_array_;
+}
+
+const std::vector<Vector_2>&
+Value::vector_2_array() const
+{
+    assert(type_ == Type::vector_2_array);
+    return *vector_2_array_;
+}
+
+const std::vector<Vector_3>&
+Value::vector_3_array() const
+{
+    assert(type_ == Type::vector_3_array);
+    return *vector_3_array_;
+}
+
+const std::vector<Matrix_3>&
+Value::matrix_3_array() const
+{
+    assert(type_ == Type::matrix_3_array);
+    return *matrix_3_array_;
+}
+
+const std::vector<Matrix_4>&
+Value::matrix_4_array() const
+{
+    assert(type_ == Type::matrix_4_array);
+    return *matrix_4_array_;
+}
+
+void
+Value::construct(const Value& x)
+{
+    type_ = x.type_;
+    switch (type_) {
+    case Type::boolean:
+        boolean_ = x.boolean_;
+        break;
+    case Type::number:
+        number_ = x.number_;
+        break;
+    case Type::vector_2:
+        new (&vector_2_)
+            std::unique_ptr<const Vector_2>(new Vector_2(*x.vector_2_));
+        break;
```

Comment:
just for my information, why do you need the second 'new' here? couldnt you just construct using the x.vector_2_ directly ?